### PR TITLE
Initial implementation of Expandable Design for Proprietary Data Exchange

### LINKED ISSUE #1734: [SDL 0083] Expandable Design for Proprietary Data Exchange
Proposal: [Expandable Design for Proprietary Data Exchange](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0083-Expandable-design-for-proprietary-data-exchange.md)

> The SystemRequest/OnSystemRequest RPCs provide a mechanism for various components within the head unit to exchange data with app(s) written specifically to communicate with those components. With the growing popularity of SDL as a means of data exchange between head units and the outside world, there is a need to provide service to more such components. Today, the list of such components is limited to the 'RequestType' enum. Expanding this list everytime would not be an ideal solution. This document proposes a new mechanism for an expandable design in such scenarios by introducing a new string field called 'RequestSubType' to be used along with a new 'RequestType' enum called 'OEM_SPECIFIC'.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/249

Steering Committee Decision:

> When discussing this proposal, there was an ask for an example of the data that would be exchanged. Ford provided the example of an embedded navigation getting traffic information from a companion app. The Steering Committee then voted to accept this proposal.

### PR DESCRIPTION
**Pull request by AByzhynar**
Fixes #1734 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
ATF test scripts will be provided

### Summary
- Made the following changes to Mobile API:
  Added new parameter "requestSubType" to SystemRequest and
  OnSystemRequest notification
  Added new RequestType - "OEM_SPECIFIC"

- Made the following changes to HMI API:
Added new parameter "requestSubType" to HMIApplication structure
and to OnAppPermissionChange notification
Added new RequestType - "OEM_SPECIFIC"

- Added handling of new parameters to policy regular
- Updated notifications triggering and their parameters handling
- Updated policy table validation and policy table update flow

### Tasks Remaining:
- [ ] [ATF test scripts]
- [ ] [Unit tests]

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/policy/policy_regular/src/policy_table/enums.cc [hunk 1]
```diff
@@ -501,6 +503,8 @@ const char* EnumToJsonString(RequestType val) {
       return "MEDIA";
     case RT_FOTA:
       return "FOTA";
+    case RT_OEM_SPECIFIC:
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/policy_table/enums.cc]: @EKuliiev Thks. Good idea. But we would like in nearest future to remove such conversions and these enums and use all stuff from API.xml

File: src/components/policy/policy_regular/src/sql_pt_representation.cc [hunk 2]
```diff
@@ -1109,15 +1123,83 @@ bool SQLPTRepresentation::SaveRequestType(
   }
 
   policy_table::RequestTypes::const_iterator it;
-  for (it = types.begin(); it != types.end(); ++it) {
+  if (!types.empty()) {
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @AKalinich-Luxoft Missed the idea.

File: src/components/policy/policy_regular/src/sql_pt_representation.cc [hunk 3]
```diff
@@ -1858,6 +1965,13 @@ bool SQLPTRepresentation::SetDefaultPolicy(const std::string& app_id) {
       !SaveRequestType(app_id, request_types)) {
     return false;
   }
+
+  policy_table::Strings request_subtypes;
+  if (!GatherRequestSubType(kDefaultId, &request_subtypes) ||
+      !SaveRequestSubType(app_id, request_subtypes)) {
+    return false;
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @AKalinich-Luxoft It does not make sense as all warning handled inside invoked methods

File: src/components/policy/policy_regular/src/sql_pt_representation.cc [hunk 4]
```diff
@@ -1006,6 +1016,10 @@ bool SQLPTRepresentation::SaveSpecificAppPolicy(
     return false;
   }
 
+  if (!SaveRequestSubType(app.first, *app.second.RequestSubType)) {
+    return false;
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @AKalinich-Luxoft Appropriate warning handled inside the method called.

File: src/components/policy/policy_regular/src/sql_pt_representation.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @AByzhynar like this:
```
if (!types.empty()) {
  LOG4CXX_WARN(logger_, "Request types not empty.");
  for (it = types.begin(); it != types.end(); ++it) {
    query.Bind(0, app_id);
    query.Bind(1, std::string(policy_table::EnumToJsonString(*it)));
    if (!query.Exec() || !query.Reset()) {
      LOG4CXX_WARN(logger_, "Incorrect insert into request types.");
      return false;
    }
  }
  return true;
}

if (types.is_initialized()) {
  LOG4CXX_WARN(logger_, "Request types empty.");
  query.Bind(0, app_id);
  query.Bind(1,
             std::string(policy_table::EnumToJsonString(
                policy_table::RequestType::RT_EMPTY)));
  if (!query.Exec() || !query.Reset()) {
    LOG4CXX_WARN(logger_, "Incorrect insert into request types.");
    return false;
  }
  return true;
}

utils::dbms::SQLQuery query_omitted(db());
if (!query_omitted.Prepare(sql_pt::kInsertOmittedRequestType)) {
  LOG4CXX_WARN(logger_, "Incorrect insert statement for request types.");
  return false;
}
LOG4CXX_WARN(logger_, "Request types omitted.");
query_omitted.Bind(0, app_id);
if (!query_omitted.Exec() || !query_omitted.Reset()) {
  LOG4CXX_WARN(logger_, "Incorrect insert into request types.");
  return false;
}

return true;
```

[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/policy/policy_regular/src/sql_pt_representation.cc]: @AKalinich-Luxoft Due to more complicated logic this idea is not applicable here.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 5]
```diff
@@ -2338,6 +2341,10 @@ bool ApplicationManagerImpl::ConvertMessageToSO(
         output[strings::params][strings::binary_data] =
             *(message.binary_data());
       }
+      Json::Value tmp;
+      namespace Formatters = NsSmartDeviceLink::NsJSONHandler::Formatters;
+      Formatters::CFormatterJsonBase::objToJsonValue(output, tmp);
+      LOG4CXX_DEBUG(logger_, "Mobile request: " << tmp.toStyledString());
```
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/src/application_manager_impl.cc]: @AKalinich-Luxoft Removed

**Pull request merged**